File: renameregex.java
import java.io.*;
import java.util.*;
import java.util.regex.*;

public class renameregex {
	public static void main(String[] args) {
		if ((args.length != 2) || args[0].length()==0 || args[1].length()==0) {
			//System.out.println(java.util.Arrays.asList(args));
			System.out.println("syntax: renameregex <src-regex> <dst-regex>");
			System.exit(1);
			return;
		}
		
		Pattern srcRegex = Pattern.compile("^" + args[0] + "$");
		String dstRegex = args[1];
		
		// Determine rename operations
		Map<String,String> renameOps = new LinkedHashMap<String,String>();
		Set<String> allDstFilenames = new HashSet<String>();
		boolean dupDstFilenames = false;
		for (String srcFilename : new File(".").list()) {
			Matcher m = srcRegex.matcher(srcFilename);
			if (m.matches()) {
				String dstFilename = m.replaceFirst(dstRegex);
				
				renameOps.put(srcFilename, dstFilename);
				if (!allDstFilenames.add(dstFilename)) {
					dupDstFilenames = true;
				}
			}
		}
		
		// Present proposed rename operations and confirm
		for (Map.Entry<String,String> curEntry : renameOps.entrySet()) {
			String srcFilename = curEntry.getKey();
			String dstFilename = curEntry.getValue();
			System.out.println(srcFilename + " => " + dstFilename);
		}
		System.out.println();
		if (dupDstFilenames) {
			System.out.println("Multiple source files would be renamed to the same destination file. Aborted.");
			return;
		}
		System.out.print("Okay? [y/N] ");
		
		String ans = new Scanner(System.in).nextLine();
		if (ans.equalsIgnoreCase("y")) {
			// Do rename operations
			for (Map.Entry<String,String> curEntry : renameOps.entrySet()) {
				String srcFilename = curEntry.getKey();
				String dstFilename = curEntry.getValue();
				
				new File(srcFilename).renameTo(new File(dstFilename));
			}
		}
	}
}

File: renameregex.py
#!/usr/bin/env python3

from optparse import OptionParser
import os
import re
import sys


def main():
    parser = OptionParser()
    parser.add_option('-r',
        action='store_true',
        dest='r',
        default=False,
        help='Looks in subdirectories recursively',
    )
    (options, args) = parser.parse_args()
    if len(args) != 2:
        parser.error('expected arguments: <filename_pattern> <replacement_pattern>')
    (filename_pattern, replacement_pattern) = args
    
    if options.r:
        def walkpaths():
            for (rootpath, dirnames, filenames) in os.walk('.', topdown=False):
                for filename in filenames:
                    yield os.path.join(rootpath, filename)
                for dirname in dirnames:
                    yield os.path.join(rootpath, dirname)
        filepaths = walkpaths()
    else:
        filepaths = os.listdir('.')
    
    # Determine rename operations
    rename_ops = []
    new_filepaths = []
    filename_pattern_re = re.compile(filename_pattern)
    for old_filepath in filepaths:
        old_filename = os.path.basename(old_filepath)
        new_filename = re.sub(filename_pattern_re, replacement_pattern, old_filename)
        if new_filename != old_filename:
            new_filepath = os.path.join(os.path.dirname(old_filepath), new_filename)
            
            rename_ops.append((old_filepath, new_filepath))
            new_filepaths.append(new_filepath)
    
    # Present proposed rename operations and confirm
    for (old_filepath, new_filepath) in rename_ops:
        print('%s => %s' % (old_filepath, new_filepath))
    print()
    if len(new_filepaths) != len(set(new_filepaths)):
        print('Multiple source files would be renamed to the same destination file. Aborted.')
        sys.exit(1)
    ok = input('Okay? [y/N] ')
    if ok.lower() != 'y':
        return
    
    # Perform rename operations
    for (old_filepath, new_filepath) in rename_ops:
        os.rename(old_filepath, new_filepath)


if __name__ == '__main__':
    main()
